Reverse string approaches (#609)
exercism/clojure: exercises/practice/reverse-string/.approaches/recursion/content.md

# Recursion

```clojure
(defn reverse-string [s]
  (loop [s s acc ""]
    (if (empty? s)
      acc
      (recur
        (rest s)
        (str (first s) acc)))))
```

## Performance considerations

> It is not necessarily bad to use recursion as in these two examples.

Reviewer: ```suggestion
It is not necessarily bad to use recursion as shown in the example above.
```

In fact, many other approaches use recursion behind the scenes as, for instance, `clojure.string/join` is implemented using recursion.

> However, we should remember that strings are immutable, so code like the examples above will be inefficient.

Reviewer: ```suggestion
However, we should remember that strings are immutable, so code like the above will be inefficient.
```

# Recursion

```clojure
(defn reverse-string [s]
  (loop [s s acc ""]
    (if (empty? s)
      acc
      (recur
        (rest s)
        (str (first s) acc)))))
```

In the community solutions you can see the distinct group of solutions using recursion to solve the exercise.

> They can look like the `loop`/`recur` example above, or perhaps use the `letfn` macro as below.

Reviewer: Then you'd have to explain why it overflows and this complicates things even more for no reason, especially when you can simply use `recur` and avoid the overflow. So, if you still want to keep the example, just use `recur` and change the name of the `str` variable since it clashes with the core function `str`. This will make the code look similar to the `loop/recur` approach.

> They can look like the `loop`/`recur` example above, or perhaps use the `letfn` macro as below.

Author: fair point about explaining further. I'll remove this example.
